Adds token authentication and usage of official API for Growatt MIN/TLX inverters

johanzander opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

This PR allows users of MIN/TLX inverters to authenticate using their API token instead of the existing username/password method. In addition, sensor data is fetched using the officially supported V1 API.

This has several benefits:
1. Using the officially supported API instead of the unofficial internal APIs is more future proof as Growatt can change the internal APIs at any time and break the integration.
2. The official API opens up for additional features not currently available, such as controlling inverter settings from Home Assistant, for example TOU settings, charge/discharge power, etc. This will be added in a following PR.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- Utilizes the V1 API for MIN/TLX devices added in growattServer 1.7.1:
https://github.com/home-assistant/core/pull/149716

- Growatt V1 API Documentation including  how to find API key: https://www.showdoc.com.cn/262556420217021/1494053950115877

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] I have followed the [perfect PR recommendations][perfect-pr]
- [X] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [X] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!

johanzander committed:
Add support for API token authentication and refactor device list retrieval logic

- Introduced API token authentication method in the config flow.
- Refactored device list retrieval to support both classic and V1 APIs.
- Enhanced error handling for API calls and improved logging.

johanzander committed:
Add support for MIN device type in Growatt integration and update sensor handling

johanzander committed:
Clarify comment regarding MIN device support in V1 API

joostlek reviewed (changes requested):
The config flow should be 100% unit tested.

Is everyone who uses growatt able to get an API key and use this API?

For the first step, check if you can use a menu, as these generally look better.

johanzander committed:
Enhance Growatt server integration: add authentication menu and unit test, improve error handling, and update UI strings for clarity

johanzander committed:
Merge branch 'dev' into feature/growatt_open_api

johanzander committed:
Merge branch 'dev' into feature/growatt_open_api

johanzander commented:
And thanks btw for proposing stepping up as codeowner for this integration @joostlek. This is my first HA contribution and still learning the project and review process so would prefer a few more PR contributions before I would be comfortable doing so :-)

johanzander committed:
Remove unreachable code for empty plants list in token authentication step

johanzander committed:
Improve error handling during initial Growatt classic API login and data retrieval

johanzander committed:
Merge branch 'dev' into feature/growatt_open_api

johanzander committed:
Updates get_device_list_classic() with improved error handling and debug

johanzander committed:
Merge branch 'dev' into feature/growatt_open_api

joostlek commented on homeassistant/components/growatt_server/__init__.py:
```diff
@@ -19,36 +20,112 @@
+        devices_dict = api.device_list(plant_id)
+    except growattServer.GrowattV1ApiError as e:
+        raise ConfigEntryError(
+            f"API error during device list: {e} (Code: {getattr(e, 'error_code', None)}, Message: {getattr(e, 'error_msg', None)})"
```
`growattServer.GrowattV1ApiError` always has `error_code` and `error_msg` as property, so we can do without `getattr`

joostlek commented on homeassistant/components/growatt_server/__init__.py:
```diff
@@ -19,36 +20,112 @@
-        plant_info = api.plant_list(user_id)
-        plant_id = plant_info["data"][0]["plantId"]
+        try:
+            plant_info = api.plant_list(user_id)
+        except Exception as ex:
+            raise ConfigEntryError(
+                f"Error communicating with Growatt API during plant list: {ex}"
+            ) from ex
+        if not plant_info or "data" not in plant_info or not plant_info["data"]:
+            raise ConfigEntryError("No plants found for this account.")
+        plant_id = plant_info["data"][0].get("plantId")
+        if not plant_id:
+            raise ConfigEntryError("Plant ID missing in plant info.")
 
     # Get a list of devices for specified plant to add sensors for.
-    devices = api.device_list(plant_id)
+    try:
+        devices = api.device_list(plant_id)
+    except Exception as ex:
+        raise ConfigEntryError(
+            f"Error communicating with Growatt API during device list: {ex}"
+        ) from ex
+
```
Why do we add these extra checks here? I think this is fine, but we need to catch more specific exceptions than just Exception

johanzander replied:
users reported uncaught issues from the library (https://github.com/home-assistant/core/issues/150732#issuecomment-3342633646) so to adress that I wanted to catch all errors I could think of, but maybe became a bit to defensive. I've done a new attempt with more specific exceptions.

joostlek commented on homeassistant/components/growatt_server/__init__.py:
```diff
@@ -19,36 +20,112 @@
+    try:
+        login_response = api.login(config[CONF_USERNAME], config[CONF_PASSWORD])
+        # DEBUG: Log the actual response structure
+    except Exception as ex:
```
Same for here, can we be more specific?

joostlek commented on homeassistant/components/growatt_server/__init__.py:
```diff
@@ -58,11 +135,24 @@ async def async_setup_entry(
-    api = growattServer.GrowattApi(add_random_user_id=True, agent_identifier=username)
-    api.server_url = url
+    # Determine API version
+    if config.get("auth_type") == "api_token":
+        api_version = "v1"
+        token = config["token"]
+        api = growattServer.OpenApiV1(token=token)
+    elif config.get("auth_type") == "password":
```
can we add constants for `"auth_type"`, `"api_token"` and `"password"`?

joostlek commented on homeassistant/components/growatt_server/config_flow.py:
```diff
@@ -65,35 +75,129 @@ async def async_step_user(
+            data_schema=data_schema,
+            errors=errors,
+            description_placeholders={
+                "note": "Token authentication only supports MIN/TLX devices. For other device types, please use username/password authentication."
```
Why is this a description placeholder? Can't we just add it to the description?

Also, would it make sense to label the `["password_auth", "token_auth"]` user facing strings with the devices they support?

johanzander replied:
and the stings are added..

joostlek commented on homeassistant/components/growatt_server/config_flow.py:
```diff
@@ -65,35 +75,129 @@ async def async_step_user(
+            # Create dictionary of plant_id -> name
+            plant_dict = {}
+            for plant in self.plants:
+                plant_id = str(plant.get("plant_id", ""))
```
why would the plant_id not be there?

johanzander replied:
dunno, only (too) defensive error handling that probably could not happen in reality. refactored to remove check and improve readability.

joostlek commented on homeassistant/components/growatt_server/coordinator.py:
```diff
@@ -67,21 +75,53 @@ def _sync_update_data(self) -> dict[str, Any]:
+                min_details = self.api.min_detail(self.device_id)
+                min_settings = self.api.min_settings(self.device_id)
+                min_energy = self.api.min_energy(self.device_id)
+                min_info = {**min_details, **min_settings, **min_energy}
+                self.data = min_info
+                _LOGGER.debug("min_info for device %s: %r", self.device_id, min_info)
```
only have things that can raise in the try block

johanzander replied:
done

joostlek commented on homeassistant/components/growatt_server/strings.json:
```diff
@@ -5,23 +5,37 @@
+        "description": "Note: API Token authentication is currently only supported for MIN/TLX devices. For other device types, please use Username & Password authentication.",
+        "menu_options": {
+          "password_auth": "Username & Password",
+          "token_auth": "API Token (MIN/TLX only)"
```
Oh never mind, you already have this description here

joostlek commented on tests/components/growatt_server/test_config_flow.py:
```diff
@@ -90,21 +163,70 @@ async def test_no_plants_on_account(hass: HomeAssistant) -> None:
+            result["flow_id"], FIXTURE_USER_INPUT_PASSWORD
+        )
+
+    assert result["type"] is FlowResultType.CREATE_ENTRY
+    assert result["data"][CONF_USERNAME] == FIXTURE_USER_INPUT_PASSWORD[CONF_USERNAME]
+    assert result["data"][CONF_PASSWORD] == FIXTURE_USER_INPUT_PASSWORD[CONF_PASSWORD]
+    assert result["data"][CONF_PLANT_ID] == "123456"
+    assert result["data"][CONF_AUTH_TYPE] == AUTH_PASSWORD
+    assert result["data"][CONF_NAME] == "Plant name"
```
can we also assert the unique id?

johanzander committed:
Address code review feedback for config flow

  - Add constants for error and abort reason strings to avoid hardcoding
  - Replace broad exception catches with specific exception types
  - Improve try/catch block structure to only wrap API calls
  - Add unique ID assertions to all config entry creation tests
  - Add test coverage for connection errors and invalid API responses
  - Achieve 100% test coverage for config_flow.py (21 tests passing)

johanzander committed:
Refactor Growatt integration to use constants for authentication types and improve error handling in data fetching

johanzander committed:
Removes defensive error handling for unrealistic edge case to improve readability where plant_id missing from plant response

johanzander committed:
Merge branch 'dev' into feature/growatt_open_api

johanzander committed:
Improve config flow tests: add recovery testing and deduplicate with parametrize

  - Update error handling tests to demonstrate flow recovery by ending in CREATE_ENTRY
  - Use pytest.mark.parametrize to deduplicate similar test patterns
  - Remove unrealistic test cases for missing plant_id scenarios
  - Tests now show complete error → retry → success flows for better resilience testing

johanzander committed:
Merge branch 'dev' into feature/growatt_open_api

johanzander committed:
Remove conditional logic from config flow tests by splitting parametrized
  tests

  Split two parametrized tests that contained if statements into separate test
   functions:
  - test_auth_no_plants → test_password_auth_no_plants +
  test_token_auth_no_plants
  - test_existing_plant_configured →
  test_password_auth_existing_plant_configured +
  test_token_auth_existing_plant_configured

joostlek reviewed (approved):
Awesome! Thanks for your patience. It would be nice if we can have someone who can give the Growatt integration more love, so be sure to add yourself as codeowner if you are willing to help us out :)

johanzander commented:
Yes, will do!
